stored_requests/multifetcher.go

```diff
@@ -0,0 +1,50 @@
+package stored_requests
+
+import (
+	"context"
+	"encoding/json"
+)
+
+// MultiFetcher is a Fetcher composed of multiple sub-Fetchers that are all polled for results.
+type MultiFetcher []Fetcher
+
+// FetchRequests implements the Fetcher interface for MultiFetcher
+func (mf *MultiFetcher) FetchRequests(ctx context.Context, ids []string) (map[string]json.RawMessage, []error) {
+	var errs []error
+	result := make(map[string]json.RawMessage, len(ids))
+	missingIDs := 0 // The number of missing IDs that each fetcher reported.
+	// If the number of errors == number of missing IDs, then it is likely that the errors are simply due
+	// to the IDs missing from the return result.
+	numIDs := len(ids)
+	// suspect fetchers ... fetchers that returned errors that don't match the number of missing IDs.
+	suspect := 0
+
+	// Loop over the fetchers
+	for _, f := range *mf {
+		remainingIDs := make([]string, 0, len(ids))
```

Comment:
You could avoid having to recreate remaining IDs each time through the loop by doing [something like this](https://github.com/prebid/prebid-server/pull/404/files#diff-c705d31b368d04ad30e87727936db388R57) to safely delete fetched ids from the slice while iterating over it, then this could be initialized outside of the loop:
```go
remainingIDs := make([]string, len(ids))
copy(remainingIDs, ids)
```

and then updated after the call to `FetchRequests` below.

Reply:
Wouldn't the copy be essentially a loop over the entries, and then we do a second loop to delete out the ones we don't want? I still think the append is more efficient, given I have sized the slice such that it will not require resizing.

Reply:
The copy would just happen once before the `for _, f := range *mf` loop; the delete loop would happen inside the `for _, f := range *mf` loop in place of the `for _, id := range ids {` loop below.

Reply:
The copy would happen once, but the internals of the copy is essentially a `for` loop (if not explicitly a `for` loop) Certainly it is not a single CPU operation, which seems to be how you are counting to say the delete would be more efficient.

Reply:
Sorry I'm not wording things clearly, I more meant that currently it's something like this:
```
for each fetcher {
    reset remaining ids to []
    for each id {
        check if id in result and if not, append to remaining ids
    }
    fetch remaining ids
}
```
which loops over all ids in each iteration of the main loop
I was suggesting something like:
```
copy ids to remaining ids
for each fetcher {
    fetch remaining ids
    for each remaining id {
        check if id in fetched ids and if so, add to result and delete from remaining ids
    }
}
```
which only loops over all ids once, then each iteration of the main loop is only looping over a shrinking slice of remaining ids (the delete is constant time).

They're both the equivalent in worst case (all ids missing from all fetchers) but the latter would perform better as each fetcher reduced the number of remaining ids. The latter also has the added benefit that `len(remainingIDs)` would be zero at the end if all IDs were filled instead of containing the remaining IDs from before the final fetcher was called.

To be honest though there's really not much performance difference in practice because we will always have a small number of fetchers (max 2) and a small list of IDs to fetch.